sw/device/lib/dif/dif_otp_ctrl.h

```diff
@@ -0,0 +1,962 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_DIF_DIF_OTP_CTRL_H_
+#define OPENTITAN_SW_DEVICE_LIB_DIF_DIF_OTP_CTRL_H_
+
+/**
+ * @file
+ * @brief <a href="/hw/ip/otp_ctrl/doc/">OTP</a> Device Interface Functions
+ */
+
+#include <stdint.h>
+
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_warn_unused_result.h"
+
+// Header Extern Guard (so header can be used from C and C++)
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+/**
+ * A toggle state: enabled, or disabled.
+ *
+ * This enum may be used instead of a `bool` when describing an enabled/disabled
+ * state.
+ */
+typedef enum dif_otp_ctrl_toggle {
+  /*
+   * The "enabled" state.
+   */
+  kDifOtpCtrlToggleEnabled,
+  /**
+   * The "disabled" state.
+   */
+  kDifOtpCtrlToggleDisabled,
+} dif_otp_ctrl_toggle_t;
+
+/**
+ * A partition within OTP memory.
+ */
+typedef enum dif_otp_ctrl_partition {
+  /**
+   * The creator software configuration area.
+   *
+   * This partition contains device-specific calibration data.
+   */
+  kDifOtpCtrlPartitionCreatorSwCfg,
+  /**
+   * The owner software configuration area.
+   *
+   * This partition contains data to e.g. enable ROM hardening features.
+   */
+  kDifOtpCtrlPartitionOwnerSwCfg,
+  /**
+   * The hardware configuration area.
+   */
+  kDifOtpCtrlPartitionHwCfg,
+  /**
+   * The device lifecycle area.
+   */
+  kDifOtpCtrlPartitionLifeCycle,
+  /**
+   * Scrambled partition 0.
+   *
+   * This paritition contains TEST lifecycle state unlock tokens.
+   */
+  kDifOtpCtrlPartitionSecret0,
+  /**
+   * Scrambled partition 1.
+   *
+   * This partition contains SRAM and flash scrambling keys.
+   */
+  kDifOtpCtrlPartitionSecret1,
+  /**
+   * Scrambled partition 2.
+   *
+   * This partition contains the RMA unlock token and the CreatorRootKey.
+   */
+  kDifOtpCtrlPartitionSecret2,
+} dif_otp_ctrl_partition_t;
```

Comment:
Any way we can statically check the numbers match up with the top instantiation?

Reply:
I don't think there's any way to do that? @msfschaffner thoughts?

Reply:
I can add a parameter to the hjson that holds the total number of partitions in the design (together with the partition offsets and sizes). Would that help?

Reply:
Not for this specific problem, but that will be extremely helpful in my implementation.